doc/development/tutorials/examples/todo.py

```diff
@@ -15,106 +24,114 @@ class todolist(nodes.General, nodes.Element):
     pass
 
 
-def visit_todo_node(self, node):
+class TodoInfo(NamedTuple):
+    docname: str
+    lineno: int
+    todo: todo
+    target: nodes.target
+
+
+def visit_todo_node(self: nodes.GenericNodeVisitor, node: nodes.Node) -> None:
     self.visit_admonition(node)
 
 
-def depart_todo_node(self, node):
+def depart_todo_node(self: nodes.GenericNodeVisitor, node: nodes.Node) -> None:
     self.depart_admonition(node)
 
 
 class TodolistDirective(Directive):
-    def run(self):
+    def run(self) -> Sequence[nodes.Node]:
         return [todolist('')]
 
 
+@contextmanager
+def get_all_todos(env: BuildEnvironment) -> Iterator[list[TodoInfo]]:
+    all_todos: list[TodoInfo] = getattr(env, 'todo_all_todos', [])
+    yield all_todos
+    env.todo_all_todos = all_todos  # type: ignore[attr-defined]
```

Comment:
```suggestion
    all_todos: list[TodoInfo] = getattr(env, 'todo_all_todos', [])
    env.todo_all_todos = all_todos  # type: ignore[attr-defined]
    yield all_todos
```

We might want to set the variable *before* it's being updated I think. Otherwise, if we do `env.todo_all_todos` before exiting the context, then we will have an AttributeError (in this example, this is not the case, but we don't know how users will use the code).

Reply:
the idea of using this pattern is that you don't ever access `env.todo_all_todos` directly.

I use this pattern in my own plugin and it works quite well. I think sphinx-needs does something similar too.

Maybe we can extend BuildEnvironment to make this pattern a little more ergonomic?
perhaps by adding a `BuildEnvironment.plugin_data` where plugin_data: `dict[str, dict[str, Any]`

Reply:
Actually, I wasn't critizing the pattern itself. It's just that if you do:

```py
with get_all_todos(env) as all_todos:
     for t in env.all_todos:
		...
```

Then you may have surprises. In addition, if you have an exception your attribute will never be set (which is not the same as guaranteeing its existence even if an exception occurs afterwards).

Reply:
> Actually, I wasn't critizing the pattern itself. It's just that if you do:
> 
> ```python
> with get_all_todos(env) as all_todos:
>      for t in env.all_todos:
> 		...
> ```

this should be:

```python
with get_all_todos(env) as all_todos:
     for t in all_todos:
	...
```

Reply:
Yes, it should but it's a tutorial that can be extended. I'm not really against it (and I would personally keep it the way you did), but I don't know if we should do it in a tutorial section (should we actually ensure that the attribute exists once you called the context manager or not? if there is an exception in between, then the attribute will not exist at all).

Reply:
i've made the attribute more 'private' and added some docs. Does that address your concerns?